### PR TITLE
[Common Lisp] basic-lists concept exercise

### LINKED ISSUE #1622: [Common Lisp] Implement new Concept Exercise: basic-lists
This issue describes how to implement the `basic-lists` concept exercise for the Common Lisp track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Learning objectives

- Be able to create a list using quote (`'`), `list`, and `cons`
- Know that lists are constructed from `cons` cells terminated by `NIL`
- Be able apply `car` / `first` and `cdr` / `rest` to get the parts of a list
- Know about `nth` and `second`, `third` ... `tenth` for accessing elements of a list
- Be able to use a list like a stack with `push` and `pop`
- Know how to use `member` to test if an item is in a list
- Be able to combine lists with `append`
- Know which functions are list-specific and which are general to sequences
- Be comfortable using basic sequence functions like `length`, `subseq`, `sort`, and `reverse`

## Out of scope

- Any higher-order functions like `mapcar` or `count-if`
- Discussion of other sequence types in lisp
- Circular, association, or property lists
- The advanced `cXr` accessors like `cadr` or `cdadar`.
- More advanced sequence functions like `search`, `substitute`, or `merge`

## Concepts

- `basic-lists`

## Prerequisites

- `cons`
- `symbols`

## Resources to refer to

- [Practical Common Lisp](http://www.gigamonkeys.com/book/they-called-it-lisp-for-a-reason-list-processing.html)

### Hints

- [Tutorialspoint](https://www.tutorialspoint.com/lisp/lisp_lists.htm)

### After

- Show that sequence functions can also be used on things like strings
- Show off some of the simpler `cXr` functions

## Representer

N/A

## Analyzer

- The analyser should be able to distinguish between the different forms of list construction (`'` vs `list` etc.)
- Be able to suggest alternatives like `car` or `first`

## Implementing

Remember that this is a basic exercise. It's important to have strong grasp on basic list manipulation and usage, but mastering all of the details is unimportant at this stage.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

### PR DESCRIPTION
**Pull request by verdammelt**
TODO:

- [x] introduction.md
- [x] (finish) instructions.md
- [x] after.md
- [x] hints.md
- [X] design.md
- [x] decide if `sort` and `reverse` are in or out
- [x] write tests
- [x] write example
- [x] write stub solution

Fixes #1622

### CODE REVIEW

**Review by TheLostLambda — COMMENTED** [review]
Overall looking great! I've left some comments and am leaning towards your second starting task (1 as opposed to 0).

For `reverse` and `sort`, I think we could defer `sort` to another exercise and I left a couple of `reverse` ideas here :)

File: languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md
Comment (verdammelt) [reply languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md:24]: def. a typo. I originally had more `n-thing` items - and decided to chop them down. Thanks for catching.

File: languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md
Comment (verdammelt) [reply languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md:7]: Yes, 0th is an affectation of mine. But also I was, wrongly, thinking about how in markdown when you have numbered items that it doesn't matter the number you put into your markdown doc - it will be rendered by a correct sequence of numbers. (Of course that is not the case here...)

Comment (TheLostLambda) [reply languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md:7]: Huh, you learn something new about markdown every day... That's pretty neat! Though I suppose they aren't expecting headers like these to be numbered

File: languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md
Comment (verdammelt) [reply languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md:11]: oops i hadn't thought it through...

Trying to think back to why I wrote this task - i think it was to have the student use `list`. So I think I want to keep it in. I'll add a prerequisite for now. But perhaps when the actual exercise is written maybe we can limit it to a number of arguments in some way.

File: languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md
Comment (verdammelt) [reply languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md:15]: But your point about the first two tasks - yes, it looks like i must have started this set of tasks twice and never noticed the duplication.

File: languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md
Comment (verdammelt) [reply languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md:37]: I liked `list-+` since it gives (to me at least) that feeling of adding, or summing, lists.  But I think I'll switch to `list-join`. Or `list-append` or even `list-prepend`.

**Comment by ErikSchierboom** [comment]
@verdammelt Is there anything I can do to help with this PR?

**Comment by verdammelt** [comment]
@ErikSchierboom I still need to write an after & hints document. But i do have a worry that this might be "too big"?

Wouldn't mind your opinion (along with @TheLostLambda of course)....

**Comment by ErikSchierboom** [comment]
@verdammelt I'll do a review later

File: languages/common-lisp/exercises/concept/basic-lists/.meta/config.json
Comment (ErikSchierboom) [inline languages/common-lisp/exercises/concept/basic-lists/.meta/config.json:9]: Empty fields can be omitted from this document.

File: languages/common-lisp/exercises/concept/basic-lists/.meta/design.md
Comment (ErikSchierboom) [inline languages/common-lisp/exercises/concept/basic-lists/.meta/design.md:72]: I think this can be removed.

File: languages/common-lisp/exercises/concept/basic-lists/.meta/design.md
Comment (ErikSchierboom) [inline languages/common-lisp/exercises/concept/basic-lists/.meta/design.md:1]: The name `basic-lists` implies are sort of hierarchy. As per [the docs](https://github.com/exercism/v3/blob/master/docs/maintainers/determining-concepts.md#naming-concepts), we don't use a hierarchy for concepts. What if you'd just name this `lists`? More advanced list exercises could then be named after the more advanced list concepts they discuss.

File: languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md
Comment (ErikSchierboom) [inline languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md:1]: While are quite some tasks here, the key thing to keep in mind that someone proficient in Common Lisp should be able to finish this exercise in 5-10 minutes. Looking at the example implementation, I think this is actually doable with the number of tasks and their difficulty. So I'm fine with how "big" this exercise is currently.

File: languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md
Comment (ErikSchierboom) [inline languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md:12]: Could you add an example of the function call to the task? See [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsinstructionsmd) and [this existing Common Lisp exercise](https://github.com/exercism/v3/blob/master/languages/common-lisp/exercises/concept/basics/.docs/instructions.md).

File: languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md
Comment (ErikSchierboom) [reply languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md:1]: I really like how the tasks are written. Very engaging!

File: languages/common-lisp/exercises/concept/basic-lists/.docs/introduction.md
Comment (ErikSchierboom) [inline languages/common-lisp/exercises/concept/basic-lists/.docs/introduction.md:1]: The introduction is currently quite long. We're aiming for [relatively concise and minimal introductions](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsintroductionmd) where the information provided should give the student just enough context to figure out the solution themselves.

What makes this introduction relatively big is that it uses a lot of code samples. Code examples should only be used to introduce new syntax (students should not need to search the web for examples of syntax). Many of the examples just show how specific functions can be used. In other words, they introduce new functions, not new syntax. As such, these examples should be omitted from the introduction (they're perfectly fine in an `after.md` document).

In many languages, the list functions are defined in a module or namespace. So for those languages, it is enough for the introduction to just explicitly mention that module which allows the student to find the relevant methods themselves (and if they can't, the hints should link to those methods). See the [F# lists exercise](https://github.com/exercism/v3/blob/master/languages/fsharp/exercises/concept/lists/.docs/introduction.md) for example.

**Review by ErikSchierboom — APPROVED** [review]
Nice! I've left some comments.

File: languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md
Comment (verdammelt) [reply languages/common-lisp/exercises/concept/basic-lists/.docs/instructions.md:1]: Thanks for the feedback! Makes me feel better about how this came out.

File: languages/common-lisp/exercises/concept/basic-lists/.meta/design.md
Comment (verdammelt) [reply languages/common-lisp/exercises/concept/basic-lists/.meta/design.md:1]: concept & exercise both renamed - thanks.

File: languages/common-lisp/exercises/concept/basic-lists/.docs/introduction.md
Comment (verdammelt) [reply languages/common-lisp/exercises/concept/basic-lists/.docs/introduction.md:1]: Looks like i had it backwards - i put lots of code examples in the introduction rather the instructions.

In Common Lisp all functions are in a single namespace: `COMMON-LISP` and the naming isn't always consistent (i.e. they don't all start with `LIST-`) for historical reasons (language was created by committee to produce a "common" lisp based on the multiple competing lisps at the time). So it is not always easy to find the function you want. The language specification is a GREAT resource as it is divided into sections such as 'characters', 'strings', 'sequences'. However it is NOT GREAT for a person new to the language.

**Comment by verdammelt** [comment]
Finalized this PR with a hints & after document.

@TheLostLambda Even though the length of this exercise got the seal of approval from Erik I want your opinion on my new idea to simplify this a bit. For example removing `member`, `subseq` and `reverse`. I think `list`, `cons`, `c[ad]r`, `append` is more than enough (along with previous exercises for quoting and equality) for a student to write Lisp code for plenty of experimentation. Hell that is most of McCarthy's language from this original paper...

Througts?

**Comment by TheLostLambda** [comment]
Hi @verdammelt ! I'd love to take a look at this soon! Today is slightly, hectic, so it may need to be tomorrow, but I'm hyped to have another exercise done :)

**Comment by verdammelt** [comment]
I'll try to push a commit today that removes some of the tasks you we can see how it would actually look. If we decide against the cut it is easy to remove the commit.

File: languages/common-lisp/exercises/concept/lists/.docs/hints.md
Comment (ErikSchierboom) [inline languages/common-lisp/exercises/concept/lists/.docs/hints.md:3]: Would it be possible to have the hints not spell out the solution (i.e. the function to use, but instead point to a resource describing the solution? See https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docshintsmd

File: languages/common-lisp/exercises/concept/lists/.docs/instructions.md
Comment (ErikSchierboom) [inline languages/common-lisp/exercises/concept/lists/.docs/instructions.md:1]: I like the shortened version!

File: languages/common-lisp/exercises/concept/lists/.docs/after.md
Comment (ErikSchierboom) [inline languages/common-lisp/exercises/concept/lists/.docs/after.md:3]: As we want the `after.md` document to contain the information that is in the `introduction.md` document, maybe expand on this a bit to show some examples of these functions?

**Review by ErikSchierboom — APPROVED** [review]

File: languages/common-lisp/exercises/concept/lists/.docs/introduction.md
Comment (TheLostLambda) [inline languages/common-lisp/exercises/concept/lists/.docs/introduction.md:17]: It looks like a couple of sneaky things are going on here. I think that the last parenthesis (of the sentence, not the code) is unclosed, and the Lisp lists have commas that they shouldn't. So `'(1, two, "III")` should just be `'(1 two "III")`. Just looks like an easy typo though :)

File: languages/common-lisp/exercises/concept/lists/.docs/introduction.md
Comment (TheLostLambda) [inline languages/common-lisp/exercises/concept/lists/.docs/introduction.md:24]: I'd probably uppercase the `two` here to `TWO`, as that's what Lisp usually spits back :)

File: languages/common-lisp/exercises/concept/lists/.meta/design.md
Comment (TheLostLambda) [inline languages/common-lisp/exercises/concept/lists/.meta/design.md:3]: I think I've been chopping out larges parts of the issues when I paste them here. I've just been leaving Learning, Scope, Concepts, and Prerequisites, I think. I can see a case for both approaches though, so whatever you think is best :)

File: languages/common-lisp/exercises/concept/lists/lists-test.lisp
Comment (TheLostLambda) [inline languages/common-lisp/exercises/concept/lists/lists-test.lisp:33]: Wow, I love this list <3

File: languages/common-lisp/exercises/concept/lists/.meta/design.md
Comment (TheLostLambda) [inline languages/common-lisp/exercises/concept/lists/.meta/design.md:21]: I don't think we have `push` or `pop` right now, and that's probably for the best, but we should move it to something like out of scope, I suppose

**Review by TheLostLambda — APPROVED** [review]
Overall looking wonderful! One typo we should definitely fix (the one with commas), but that's really the only major thing :) Thanks for putting this all together, it looks great!

File: languages/common-lisp/exercises/concept/lists/.docs/instructions.md
Comment (verdammelt) [reply languages/common-lisp/exercises/concept/lists/.docs/instructions.md:1]: Thank you for this comment - not for the encouragement and positive feedback - but for putting it here where I can see I failed to edit something!!!!

File: languages/common-lisp/exercises/concept/lists/.docs/after.md
Comment (verdammelt) [reply languages/common-lisp/exercises/concept/lists/.docs/after.md:13]: Good point.

File: languages/common-lisp/exercises/concept/lists/.docs/introduction.md
Comment (verdammelt) [reply languages/common-lisp/exercises/concept/lists/.docs/introduction.md:7]: good point, I'll switch to either `TWO` or `:TWO`

File: languages/common-lisp/exercises/concept/lists/.docs/introduction.md
Comment (verdammelt) [reply languages/common-lisp/exercises/concept/lists/.docs/introduction.md:17]: agh! also note the unclosed paren!

File: languages/common-lisp/exercises/concept/lists/lists-test.lisp
Comment (verdammelt) [reply languages/common-lisp/exercises/concept/lists/lists-test.lisp:33]: I used up all my creativity on this one. I especially like "right handed macaroni"

**Comment by ErikSchierboom** [comment]
Great to have this merged! Thanks @verdammelt

**Pull request merged**
